### PR TITLE
Add test cases to test sata and nvme disks hot-add/remove

### LINKED ISSUE #61: [New Feature] Support to test sata and nvme disks hot-add/remove on Linux
##### SUMMARY

Presents cases support pvscsi/lsilogic/lsilogicsas disks hot-add and remove. We need to add new cases to support sata and nvme disks testing.

### PR DESCRIPTION
**Pull request by keirazhang**
Fix #61

### CODE REVIEW

**Comment by keirazhang** [comment]
```
Testbed information:
+-----------------------------------------------+
| Product | Version | Build    | Hostname or IP |
+-----------------------------------------------+
| vCenter | 7.0.2   | 17694817 | 10.162.69.226  |
+-----------------------------------------------+
| ESXi    | 7.0.2   | 17630552 | 10.162.68.177  |
+-----------------------------------------------+

VM information:
+-----------------------------------------------------+
| VM Name             | test_vm                       |
+-----------------------------------------------------+
| VM IP               | 10.162.95.223                 |
+-----------------------------------------------------+
| Guest OS Type       | SLES 15.3 x86_64              |
+-----------------------------------------------------+
| VM Tools            | 11.2.5.26209 (build-17337674) |
+-----------------------------------------------------+
| Guest ID            | sles15_64Guest                |
+-----------------------------------------------------+
| Hardware Version    | vmx-19                        |
+-----------------------------------------------------+

Test Results (Total: 5, Failed: 0, No Run: 0, Elapsed Time: 00:24:40):
+--------------------------------------------------+
| Name                        | Status | Exec Time |
+--------------------------------------------------+
| paravirtual_vhba_device_ops | Passed | 00:07:40  |
| lsilogic_vhba_device_ops    | Passed | 00:04:20  |
| lsilogicsas_vhba_device_ops | Passed | 00:03:54  |
| sata_vhba_device_ops        | Passed | 00:03:54  |
| nvme_vhba_device_ops        | Passed | 00:03:45  |
+--------------------------------------------------+
```

**Comment by keirazhang** [comment]
```
Testbed information:
+-----------------------------------------------+
| Product | Version | Build    | Hostname or IP |
+-----------------------------------------------+
| vCenter | 7.0.2   | 17694817 | 10.187.110.126 |
+-----------------------------------------------+
| ESXi    | 7.0.2   | 17630552 | 10.187.106.229 |
+-----------------------------------------------+

VM information:
+-----------------------------------------------------+
| VM Name             | test_vm                       |
+-----------------------------------------------------+
| VM IP               | 10.187.103.8                  |
+-----------------------------------------------------+
| Guest OS Type       | CentOS 8.4 x86_64             |
+-----------------------------------------------------+
| VM Tools            | 11.2.0.23855 (build-16938113) |
+-----------------------------------------------------+
| Guest ID            | centos8_64Guest               |
+-----------------------------------------------------+
| Hardware Version    | vmx-19                        |
+-----------------------------------------------------+

Test Results (Total: 5, Failed: 0, No Run: 2, Elapsed Time: 00:07:45):
+----------------------------------------------------+
| Name                        |   Status | Exec Time |
+----------------------------------------------------+
| paravirtual_vhba_device_ops |   Passed | 00:02:55  |
| lsilogic_vhba_device_ops    | * No Run | 00:00:18  |
| lsilogicsas_vhba_device_ops | * No Run | 00:00:16  |
| sata_vhba_device_ops        |   Passed | 00:02:01  |
| nvme_vhba_device_ops        |   Passed | 00:01:54  |
+----------------------------------------------------+
```

**Comment by keirazhang** [comment]
```
Testbed information:
+---------------------------------------------------------+
| Product | Version | Build    | Hostname or IP           |
+---------------------------------------------------------+
| ESXi    | 7.0.2   | 17630552 | cpdbj-214.eng.vmware.com |
+---------------------------------------------------------+

VM information:
+-----------------------------------------------------+
| VM Name             | qiz-ubuntu-21.04-0610         |
+-----------------------------------------------------+
| VM IP               | 10.117.16.91                  |
+-----------------------------------------------------+
| Guest OS Type       | Ubuntu 21.04 x86_64           |
+-----------------------------------------------------+
| VM Tools            | 11.2.5.26209 (build-17337674) |
+-----------------------------------------------------+
| Guest ID            | ubuntu64Guest                 |
+-----------------------------------------------------+
| Hardware Version    | vmx-19                        |
+-----------------------------------------------------+

Test Results (Total: 5, Failed: 0, No Run: 0, Elapsed Time: 00:26:56):
+--------------------------------------------------+
| Name                        | Status | Exec Time |
+--------------------------------------------------+
| paravirtual_vhba_device_ops | Passed | 00:04:04  |
| lsilogic_vhba_device_ops    | Passed | 00:07:46  |
| lsilogicsas_vhba_device_ops | Passed | 00:07:52  |
| sata_vhba_device_ops        | Passed | 00:03:34  |
| nvme_vhba_device_ops        | Passed | 00:03:19  |
+--------------------------------------------------+
```

File: common/vm_get_devices_with_ctrl_key.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Get VM device with the device key
```
Comment (Tomorrow9) [inline common/vm_get_devices_with_ctrl_key.yml:4]: I think this description can be more precise about devices with the same controller.

File: common/vm_hot_add_remove_disk_ctrl.yml [hunk 2]
```diff
@@ -23,6 +23,12 @@
         type: "{{ disk_controller_type }}"
         controller_number: "{{ disk_controller_number | default(omit) }}"
   register: disk_controller_facts
+  until:
```
Comment (Tomorrow9) [inline common/vm_hot_add_remove_disk_ctrl.yml:26]: As before we do not use retry when setting configure, usually use retry for getting configure info.

File: linux/vhba_hot_add_remove/get_new_disk_controller_num.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+- name: "Loop to get new {{ new_ctrl_type }} controller number"
+  set_fact:
+    new_ctrl_number: "0"
+    new_unit_number: "0"
+    existing_ctrl_numbers: "{{ disk_controller_facts['disk_controller_data'][new_ctrl_type].values() | map(attribute='controller_busnumber') | list }}"
+
+- debug: var=disk_controller_facts['disk_controller_data'][new_ctrl_type]
```
Comment (Tomorrow9) [inline linux/vhba_hot_add_remove/get_new_disk_controller_num.yml:10]: Missing a "name" for this debug.

File: linux/vhba_hot_add_remove/sata_vhba_device_ops.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Description:
+#   This test case is used for check hotadd and hot remove PVSCSI
```
Comment (Tomorrow9) [inline linux/vhba_hot_add_remove/sata_vhba_device_ops.yml:5]: The controller type in the description need to be updated.

File: linux/vhba_hot_add_remove/nvme_vhba_device_ops.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Description:
+#   This test case is used for check hotadd and hot remove PVSCSI
```
Comment (Tomorrow9) [inline linux/vhba_hot_add_remove/nvme_vhba_device_ops.yml:5]: The controller type in the description need to be updated.

File: linux/vhba_hot_add_remove/get_new_disk_node_existing_ctrl.yml [hunk 6]
```diff
@@ -3,50 +3,69 @@
 ---
 - name: "Set fact of SCSI controller type name mapping"
   set_fact:
-    scsi_name_map:
+    controller_name_map:
```
Comment (Tomorrow9) [inline linux/vhba_hot_add_remove/get_new_disk_node_existing_ctrl.yml:6]: "SCSI" in "name" part can also be changed to disk controller type.

File: linux/vhba_hot_add_remove/vhba_device_hot_add_remove.yml [hunk 7]
```diff
@@ -2,107 +2,106 @@
...
-      vars:
-        vmtools_check: "False"
-
     - name: "Set fact of the iozone file path"
       set_fact:
         iozone_tools_path: "{{ '../../tools/iozone_64bit' if guest_os_ansible_architecture == 'x86_64' else '../../tools/iozone_32bit' }}"
     - debug:
         msg: "Will use iozone tool to test new added disk: {{ iozone_tools_path }}"
 
+    - name: "The controller type for new disk is sata controller"
```
Comment (Tomorrow9) [inline linux/vhba_hot_add_remove/vhba_device_hot_add_remove.yml:26]: Typo of "sata" here.

File: linux/vhba_hot_add_remove/get_new_disk_node_existing_ctrl.yml [hunk 8]
```diff
@@ -3,50 +3,69 @@
...
+            existing_disks: "{{ device_list_with_ctrl_key }}"
+
+        - name: "Get all occupied disk unit numbers"
+          set_fact:
+            existing_disk_unit_numbers: "{{ existing_disks | map(attribute='unitNumber') | list }}"
+      when: new_disk_ctrl_type == 'sata'
+
+    - name: "Set the new disk controller unit number"
+      set_fact:
+        new_unit_number: "{{ range(0, 16) | difference(existing_disk_unit_numbers) | min }}"
```
Comment (Tomorrow9) [inline linux/vhba_hot_add_remove/get_new_disk_node_existing_ctrl.yml:65]: For SATA and NVMe controller, it has different disk unit limitation other than 16.

File: linux/vhba_hot_add_remove/vhba_device_hot_add_remove.yml [hunk 9]
```diff
@@ -2,107 +2,106 @@
...
+        disk_controller_facts_old: "{{ disk_controller_facts['disk_controller_data'][new_ctrl_type] }}"
     - name: "Get existing SCSI controller number"
       set_fact:
-        total_existing_scsi_ctrls: "{{ disk_controller_facts['disk_controller_data']['scsi'] | length }}"
+        total_existing_controller_num: "{{ disk_controller_facts['disk_controller_data'][new_ctrl_type] | length }}"
     - name: "Print existing SCSI controller number in VM '{{ vm_name }}'"
-      debug: var=total_existing_scsi_ctrls
+      debug: var=total_existing_controller_num
+
+    - block:
```
Comment (Tomorrow9) [inline linux/vhba_hot_add_remove/vhba_device_hot_add_remove.yml:52]: If there is already 4 controllers with the same type, we can also do hotadd disk test?

File: linux/vhba_hot_add_remove/get_new_disk_controller_num.yml (hunk 3 above)
Comment (Tomorrow9) [reply linux/vhba_hot_add_remove/get_new_disk_controller_num.yml:10]: We have a common task `common/vm_get_new_vhba_bus_number.yml` to do this.

File: linux/vhba_hot_add_remove/get_new_disk_node_existing_ctrl.yml (hunk 8 above)
Comment (Tomorrow9) [reply linux/vhba_hot_add_remove/get_new_disk_node_existing_ctrl.yml:65]: We have a common task `common/vm_get_new_disk_node_info.yml ` to do this.

Comment (keirazhang) [reply linux/vhba_hot_add_remove/get_new_disk_node_existing_ctrl.yml]: We will only cover new disk on new controller, so the new unit number will be 0.

**Comment by keirazhang** [comment]
Update NVMe test scenario. We will only perform NVMe controller and disk hot add/remove testing on VM with hardware version of vmx-19 and later.
```
Testbed information:
+-----------------------------------------------+
| Product | Version | Build    | Hostname or IP |
+-----------------------------------------------+
| vCenter | 7.0.0   | 15952599 | 10.186.199.62  |
+-----------------------------------------------+
| ESXi    | 7.0.0   | 15843807 | 10.186.192.99  |
+-----------------------------------------------+

VM information:
+-----------------------------------------------------+
| VM Name             | test_vm                       |
+-----------------------------------------------------+
| VM IP               | 10.186.199.119                |
+-----------------------------------------------------+
| Guest OS Type       | CentOS 8.4 x86_64             |
+-----------------------------------------------------+
| VM Tools            | 11.2.0.23855 (build-16938113) |
+-----------------------------------------------------+
| Guest ID            | centos8_64Guest               |
+-----------------------------------------------------+
| Hardware Version    | vmx-17                        |
+-----------------------------------------------------+

Test Results (Total: 6, Failed: 0, No Run: 4, Elapsed Time: 00:04:32):
+----------------------------------------------------+
| Name                        |   Status | Exec Time |
+----------------------------------------------------+
| deploy_vm                   | * No Run | 00:00:01  |
| paravirtual_vhba_device_ops |   Passed | 00:02:03  |
| lsilogic_vhba_device_ops    | * No Run | 00:00:17  |
| lsilogicsas_vhba_device_ops | * No Run | 00:00:22  |
| sata_vhba_device_ops        |   Passed | 00:01:18  |
| nvme_vhba_device_ops        | * No Run | 00:00:10  |
+----------------------------------------------------+
```

**Review by Tomorrow9 — APPROVED** [review]
LGTM, thanks.

**Pull request merged**